src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,436 @@
+import time
+from datetime import datetime, timedelta
+
+import msal
+import pandas as pd
+import requests
+from dateutil import parser
+from pytz import timezone, utc
+
+from .PowerBiClient import PowerBiClient
+
+
+class PowerBi:
+    def __init__(
+        self,
+        client: PowerBiClient,
+        *,
+        workspace_id: str = None,
+        workspace_name: str = None,
+        dataset_id: str = None,
+        dataset_name: str = None,
+        max_minutes_after_last_refresh: int = 60 * 12,
+        timeout_in_seconds: int = 60 * 15,
+        local_timezone_name: str = "Europe/Copenhagen",
+        ignore_errors: bool = False,
+    ):
+        """
+        Allows refreshing PowerBI datasets and checking if the last refresh
+            completed successfully.
+        If no workspace is specified, a list of available workspaces will be displayed.
+        If no dataset is specified, a list of available datasets in the given workspace
+            will be displayed.
+
+        :param PowerBiClient client: PowerBI client credentials.
+
+        :param str workspace_id: The GUID of the workspace.
+        :param str workspace_name: The name of the workspace
+            (specified instead of the workspace_id).
+        :param str dataset_id: The GUID of the dataset.
+        :param str dataset_name: The name of the dataset
+            (specified instead of the dataset_id).
+        :param int max_minutes_after_last_refresh: The number of minutes
+            for which the last succeeded refresh is considered valid,
+            or 0 to disable time checking. Default is 12 hours.
+        :param bool timeout_in_seconds: The number of seconds after which
+            the refresh() method times out. Default is 15 minutes.
+        :param str local_timezone_name: The time zone to use when showing
+            refresh timestamps.
+        :param bool ignore_errors: True to print errors in the output
+            or False (default) to cast an exception.
+        """
+
+        if workspace_id is not None and workspace_name is not None:
+            raise ValueError("Specify either 'workspace_id' or 'workspace_name'!")
+        if dataset_id is not None and dataset_name is not None:
+            raise ValueError("Specify either 'dataset_id' or 'dataset_name'!")
+
+        self.workspace_id = workspace_id
+        self.workspace_name = workspace_name
+        self.dataset_id = dataset_id
+        self.dataset_name = dataset_name
+
+        # Set access parameters
+        self.client = client
+
+        self.max_minutes_after_last_refresh = max_minutes_after_last_refresh
+        self.min_refresh_time_utc = datetime.now(utc) - timedelta(
```

Comment:
Shouldn't this be calculated in the _verify_last_refresh function. This will be wrong if the class is created in a long running process?

Reply:
Agreed.